samples/algorithms/sudoku-grover/Sudoku.qs

```diff
@@ -143,21 +137,102 @@ namespace Microsoft.Quantum.Samples.SudokuGrover {
         }
     }
 
+
+    /// # Summary
+    /// Encodes stating number constraints into amplitudes.
+    ///
+    /// # Inputs
+    /// ## nVertices
+    /// The number of vertices in the graph.
+    /// ## bitsPerColor
+    /// The bit width for number of colors.
+    /// ## startingNumberConstraints
+    /// The array of (Vertex#, Color) specifying the disallowed colors for vertices.
+    ///
+    /// # Output
+    /// A 2D array of amplitudes where the first index is the cell and the second index is the value of a basis state (i.e., value) for the cell. =
+    /// Allowed amplitudes will have a value 1.0, disallowed amplitudes 0.0
+    function AllowedAmplitudes(
+        nVertices : Int,
+        bitsPerColor : Int,
+        startingNumberConstraints : (Int, Int)[]
+    ) : Double[][] {
+        mutable amplitudes = [[1.0, size=1 <<< bitsPerColor], size=nVertices];
+        for (cell, value) in startingNumberConstraints {
+            set amplitudes w/= cell <- (amplitudes[cell] w/ value <- 0.0);
+        }
+        return amplitudes;
+    }
+
+
+    /// # Summary
+    /// Prepare an equal superposition of all basis states that satisfy the constraints
+    /// imposed by the digits already placed in the grid.
+    ///
+    /// # Inputs
+    /// ## nVertices
+    /// The number of vertices in the graph.
+    /// ## bitsPerColor
+    /// The bit width for number of colors.
+    /// ## startingNumberConstraints
+    /// The array of (Vertex#, Color) specifying the disallowed colors for vertices.
+    ///
+    /// # Remarks
+    /// Creates the search space. Using the allowed amplitudes prepares uniform superposition of all allowed values for each cell
+    operation PrepareSearchStatesSuperposition(
+        nVertices : Int,
+        bitsPerColor : Int,
+        startingNumberConstraints : (Int, Int)[],
+        register : Qubit[]
+    ) : Unit is Adj + Ctl {
+        // Split the given register into nVertices chunks of size bitsPerColor.
+        let colorRegisters = Chunks(bitsPerColor, register);
+        // For each vertex, create an array of possible states we're looking at.
+        let amplitudes = AllowedAmplitudes(nVertices, bitsPerColor, startingNumberConstraints);
+        // For each vertex, prepare a superposition of its possible states on the chunk storing its color.
+        for (amps, chunk) in Zipped(amplitudes, colorRegisters) {
+            PrepareArbitraryStateD(amps, LittleEndian(chunk));
```

Comment:
This could be a relatively expensive operation, depending. If gate depth is getting to be unwieldy, it may be good to investigate replacing this by something that prepares the state more directly, using your prior knowledge about the structure of that state.

Reply:
I don't think this is a big issue, as on simulators adding a bit of depth doesn't have a major impact and on hardware we are only able to run very limited instances where the state prep has not caused issues in practice